### PR TITLE
Configure Telegraf by environment variables

### PR DESCRIPTION
**Pull request by devkid**
### Required for all PRs:

- [ ] CHANGELOG.md updated (we recommend not updating this until the PR has been approved by a maintainer)
- [X] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)

This implements #1383. Telegraf can now be configured via environment
variables that start with TELEGRAF_. See docs/CONFIGURATION.md for the
updated documentation.

### CODE REVIEW

**Comment by envintus** [comment]
This would make running telegraf as a k8s daemonset so much easier.

**Comment by johnrengelman** [comment]
would it be worthwhile to consider/possible to look at switching to something like https://github.com/spf13/viper?

**Comment by devkid** [comment]
@johnrengelman (Without thorough testing and just by looking at the README) It seems like viper doesn't support iterating over top level items (tables) or accessing [arrays of tables](https://github.com/toml-lang/toml#array-of-tables) in TOML. I think using viper would require to break the config format.

**Comment by xuqingfeng** [comment]
Should `TELEGRAF_OUTPUTS_INFLUXDB_URL` (https://github.com/influxdata/telegraf/pull/2136/files#diff-8e779ee70edd4f78e5043b4b2ce31d68R188) be `TELEGRAF_OUTPUTS_INFLUXDB_0_URL_0` ?

**Comment by devkid** [comment]
@xuqingfeng `TELEGRAF_OUTPUTS_INFLUXDB_URL` would be the same as `TELEGRAF_OUTPUTS_INFLUXDB_0_URL` (see [here](https://github.com/influxdata/telegraf/pull/2136/files#diff-8e779ee70edd4f78e5043b4b2ce31d68R345)). The `_0` at the end will not work; if you want to provide multiple URLs, you would need to pass a TOML list, e.g. `TELEGRAF_OUTPUTS_INFLUXDB_URL='["somehost", "someotherhost"]'`. The notation with indices is only needed for specifying multiple top-level sections, e.g. multiple InfluxDB output configurations. The `url` setting in the InfluxDB output plugin is for [backwards compatibility](https://github.com/influxdata/telegraf/blob/master/plugins/outputs/influxdb/influxdb.go#L21) and is only a string, not a list.

File: internal/config/config_test.go [hunk 1]
```diff
@@ -174,3 +174,60 @@ func TestConfig_LoadDirectory(t *testing.T) {
...
+	}
+	assert.NoError(t, filter.Compile())
+	mConfig := &models.InputConfig{
+		Name:     "memcached",
+		Filter:   filter,
+		Interval: 5 * time.Second,
+	}
+	mConfig.Tags = make(map[string]string)
+
+    c.LoadEnvConfig()
```
Comment (Mic92) [inline internal/config/config_test.go]: Indentation

**Comment by devkid** [comment]
Question to maintainers: are you ok with the way I implemented parsing of the environment variables, especially [here](https://github.com/devkid/telegraf/blob/769327dafc181f0056671eda19d28181f6559881/internal/config/config.go#L690)? I would have liked it better to parse only the values of the given environment variables but in the used TOML library there seems to be no exposed function that does that, only toml.Parse which requires the `key = "value"` format; same goes for config subsections.

**Comment by jforman** [comment]
I'm interested in using this functionality but curious when it will be in a released build of Telegraf. Since there are no milestones listed (and 1.5.2 is currently the stable version), when should we expect this support (config specified via environment variables) to land in a numbered Telegraf build? Thanks!

**Comment by danielnelson** [comment]
We are not currently planning to include this in mainline Telegraf until after we rework the config system, work which itself is on hold for now.  You will need to compile it yourself for the time being.  

Keep in mind you can use the existing [environment variable replacement](https://github.com/influxdata/telegraf/blob/master/docs/CONFIGURATION.md#environment-variables) to get some of the benefits of this patch.

**Comment by manish-panwar** [comment]
I am thinking if we should use `TELEGRAF_` as indicator, or ${env:my_env_name}?

**Pull request closed without merging**